### PR TITLE
Provide pid in imports callback context

### PR DESCRIPTION
**Pull request by ionull**
Then we can call exported functions in imported functions.

### CODE REVIEW

**Comment by ionull** [comment]
When I call the exported function in the imports function callback, there is a timeout exception. Any idea?

**Comment by tessi** [comment]
hey @ionull - I believe I know where the timeout comes from. My suspicion is that this is a rust-side mutex the code it waiting on indefinitely that causes the timeout in elixir-land.

What happens when rust calls into an elixir function is that we have an unlocked mutex on the `store_or_caller` (the place we all wasm state is stored). That mutex stays locked during the call. Now when you, in elixir-land, call back into rust we need to access that same `store_or_caller` to mutate the wasm state, trying to unlock that same mutex. But it will never unlock, because we're still running the imported elixir-function.

**Comment by ionull** [comment]
Hi @tessi , the problem is that GenServer has not replied, and the status is locked on the GenServer side, where you can't call with pid, but we still can use the instance to call the wasm exported function like this: 
`
Wasmex.Instance.call_exported_function(store, instance, "__demo_export_0", 
			[
				a,
				b
			],
			:from
			) 

		receive do
			{:returned_function_call, {:ok, [result]}, :from} -> :ok
			result
		after
			1000 -> 
				raise "timeout on exported function call"
		end
`
It's quite useful if we can get pid and instance from callback context.

**Comment by tessi** [comment]
uhh, interesting. so I was wrong. Nice to learn something :) I get why this is useful. Could you add a test for this scenario, please, so we can make sure not to break it in the future?

**Comment by ionull** [comment]
I tried to run the test, but I am not familiar with the rustler precompiled:

== Compilation error in file lib/wasmex/native.ex ==
** (RuntimeError) Error while downloading precompiled NIF: the precompiled NIF file does not exist in the checksum file. Please consider run: `mix rustler_precompiled.download Wasmex.Native --only-local` to generate the checksum file..

You can force the project to build from scratch with:

    config :rustler_precompiled, :force_build, wasmex: true

In order to force the build, you also need to add Rustler as a dependency in your `mix.exs`:

    {:rustler, ">= 0.0.0", optional: true}

    lib/wasmex/native.ex:8: (module)

**Comment by ionull** [comment]
What should I run or setup before `mix test`?

**Comment by ionull** [comment]
@tessi Hi, test case done.

**Comment by tessi** [comment]
> @tessi Hi, test case done.

perfect! the case looks great to me. thanks a lot. let's see what CI says and merge when/if it's green :)

File: test/wasmex_test.exs [hunk 1]
```diff
@@ -427,4 +427,40 @@ defmodule WasmexTest do
...
+          local.get 0
+          local.get 1
+          i32.add
+        )
+      )
+      """
+
+      imports = %{
+        env: %{add_import: {:fn, [:i32, :i32], [:i32], fn %{instance: instance, caller: caller}, a, b -> 
+          Wasmex.Instance.call_exported_function(caller, instance, "call_add", [a,b], :from)
```
Comment (tessi) [inline test/wasmex_test.exs:451]:
```suggestion
          Wasmex.Instance.call_exported_function(caller, instance, "call_add", [a, b], :from) 
```

the formatter seems to be nitpicking about a whitespace here

**Comment by ionull** [comment]
Is there anything else that I should do?

**Comment by tessi** [comment]
I wanted to say no, but the formatter is still upset. can you make sure `mix format --check-formatted` works? probably running `mix format` and a new commit is enough. Other than making formatters happy, I'm happy to merge :)

**Comment by ionull** [comment]
@tessi oh, I see. Have done the `mix format`, please check

**Comment by tessi** [comment]
> @tessi oh, I see. Have done the `mix format`, please check

Thank! And thanks again for your work 💚

**Pull request merged**